dot/parachain/collator-protocol/validator_side.go

```diff
@@ -115,12 +122,323 @@ func (cpvs CollatorProtocolValidatorSide) handleNetworkEvents(event network.Netw
+
+	// 1. Find out removed leaves (hashes) and newly added leaves
+	// 2. Go over each new leaves,
+	// - check if perspective parachain mode is enabled
+	// - assign incoming
+	// - insert active leaves and per relay parent
+	activeLeaves := cpvs.activeLeaves
+
+	removed := []common.Hash{}
+	for activeLeaf := range activeLeaves {
+		if !slices.Contains(view.heads, activeLeaf) {
+			removed = append(removed, activeLeaf)
+		}
+	}
+
+	newlyAdded := []common.Hash{}
+	for _, head := range view.heads {
+		if _, ok := activeLeaves[head]; !ok {
+			newlyAdded = append(newlyAdded, head)
+		}
+	}
+
+	// handled newly added leaves
+	for _, leaf := range newlyAdded {
+		mode := perspectiveParchainMode(leaf)
+
+		perRelayParent := &PerRelayParent{
+			prospectiveParachainMode: mode,
+		}
+
+		err := cpvs.assignIncoming(leaf, perRelayParent)
+		if err != nil {
+			return fmt.Errorf("assigning incoming: %w", err)
+		}
+		cpvs.activeLeaves[leaf] = mode
+		cpvs.perRelayParent[leaf] = *perRelayParent
+
+		if mode.IsEnabled {
+			// TODO: Add it when we have async backing
+			// https://github.com/paritytech/polkadot-sdk/blob/aa68ea58f389c2aa4eefab4bf7bc7b787dd56580/polkadot/node/network/collator-protocol/src/validator_side/mod.rs#L1303
+		}
+	}
+
+	// handle removed leaves
+	for _, leaf := range removed {
+		delete(cpvs.activeLeaves, leaf)
+
+		mode := perspectiveParchainMode(leaf)
+		pruned := []common.Hash{}
+		if mode.IsEnabled {
+			// TODO Do this when we have async backing
+			// https://github.com/paritytech/polkadot-sdk/blob/aa68ea58f389c2aa4eefab4bf7bc7b787dd56580/polkadot/node/network/collator-protocol/src/validator_side/mod.rs#L1340
+		} else {
+			pruned = append(pruned, leaf)
+		}
+
+		for _, prunedLeaf := range pruned {
+			perRelayParent, ok := cpvs.perRelayParent[prunedLeaf]
+			if ok {
+				cpvs.removeOngoing(perRelayParent)
+				delete(cpvs.perRelayParent, prunedLeaf)
+			}
+
+			for fetchedCandidateStr := range cpvs.fetchedCandidates {
+				fetchedCollation := fetchedCollationInfo{}
+				err := fetchedCollation.FromString(fetchedCandidateStr)
+				if err != nil {
+					// this should never really happen
+					return fmt.Errorf("getting fetched collation from string: %w", err)
+				}
+
+				if fetchedCollation.relayParent == prunedLeaf {
+					delete(cpvs.fetchedCandidates, fetchedCandidateStr)
+				}
+			}
+		}
+
+		// TODO
+		// Remove blocked advertisements that left the view. cpvs.BlockedAdvertisements
+		// Re-trigger previously failed requests again. requestUnBlockedCollations
+		// prune old advertisements
+		// https://github.com/paritytech/polkadot-sdk/blob/aa68ea58f389c2aa4eefab4bf7bc7b787dd56580/polkadot/node/network/collator-protocol/src/validator_side/mod.rs#L1361-L1396
+
+	}
+
+	return nil
+}
+
+func (cpvs *CollatorProtocolValidatorSide) removeOngoing(perRelayParent PerRelayParent) {
+	if perRelayParent.assignment != nil {
+		entry := cpvs.currentAssignments[*perRelayParent.assignment]
+		entry--
+		if entry == 0 {
```

Comment:
negative values are allowed here?

Reply:
no, entry should get removed on 0. edited this as well